File: 207-CourseSchedule/207-CourseSchedule.py
# Last updated: 8/16/2026, 9:51:21 PM
from collections import defaultdict, deque

class Solution:
    # def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
    #     # detect DAG, Directed acyclic graph
    #     # DFS

    #     # graph[preq] = [list of courses can be taken after preq]
    #     graph = defaultdict(list)
        
    #     for course, preq in prerequisites:
    #         graph[preq].append(course)
        
    #     # 0: unvisited, 1: visiting, 2: visited
    #     state = [0] * numCourses

    #     # for current course 'c', can we finish the learning of 'c'
    #     def dfs(c: int) -> bool:
    #         # dfs exit
    #         if state[c] == 2: # processed before, no cyle found
    #             return True
    #         if state[c] == 1: # c is in current DFS path, found cycle
    #             return False
            
    #         # mark as visiting and check its neighbors 
    #         state[c] = 1
    #         for c_next in graph[c]:
    #             if not dfs(c_next):
    #                 return False
    #         # mark as visited
    #         state[c] = 2            
    #         return True

    #     for c in range(numCourses):
    #         if not dfs(c):
    #             return False
        
    #     return True


    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        # BFS
        # process courses without prerequsiites first

        graph = defaultdict(list)
        indegree = [0] * numCourses

        for course, preq in prerequisites:
            graph[preq].append(course)
            indegree[course] += 1
        
        # Add all courses with no prerequisites (indegree == 0) to the queue
        queue = deque([c for c in range(numCourses) if indegree[c] == 0])

        finished = 0 # the total number of courses successfully processed
        while queue:
            node = queue.popleft()
            finished += 1
            # decrease the indegree of all downstream neighbor courses
            for neighbor in graph[node]:
                indegree[neighbor] -= 1
                # if a neighbor's prerequsites are fully satisfied, add it to the queue
                if indegree[neighbor] == 0:
                    queue.append(neighbor)
        
        # if we successfully proceed all courses, no cycle exists
        return finished == numCourses
            

File: 207-CourseSchedule/207-CourseSchedule.java
// Last updated: 8/16/2026, 9:51:24 PM
class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        // if there is a cycle in directed graph
        // topological sort via DFS
        boolean[] marked = new boolean[numCourses];
        // graph[i] = all adjacent nodes / prerequisites
        List<List<Integer>> graph = new ArrayList<>();
        for(int i = 0; i < numCourses; i ++) graph.add(new ArrayList<Integer>());

        for(int[] pair: prerequisites) {
            graph.get(pair[0]).add(pair[1]);
        }
        System.out.println(graph);

        // status: 0-unknown, 1-visiting, 2-visited
        int[] status = new int[numCourses];

        for(int i = 0; i < numCourses; i ++) {
            if(!marked[i]) {
                if(!dfs(graph, status, i)) return false;
                // if dfs return false, means cycle found, return false

                marked[i] = true;
            }
        }
        return true;
    }

    private boolean dfs(List<List<Integer>> graph, int[] status, int i) {
        if(status[i] == 1) return false; // meet visiting nodes, has cycle, return false
        if(status[i] == 2) return true; // already visited this node-i before, no cycle

        // mark node-i as visiting
        status[i] = 1;
        // iterate node-i adjacent nodes
        for(int j: graph.get(i)) {
            if( !dfs(graph, status, j) ) return false;
            // if found cycle, return false
        }

        // finish checking, mark node-i as visited
        status[i] = 2;
        return true;
    }
}